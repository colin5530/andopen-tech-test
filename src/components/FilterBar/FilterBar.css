.c-FilterBar {
  display: inline-flex;
  width: 100%;
}

.c-FilterBar-filter-title {
  font-weight: 700;
}

.c-FilterBar-slider-container {
  flex-grow: 3;
  margin: 2px;
  padding: 10px;
}

.c-FilterBar-region-container {
  flex-grow: 1;
  margin: 2px;
  padding: 10px;
}

.c-FilterBar-gender-container {
  flex-grow: 2;
  margin: 2px;
  padding: 10px;
}