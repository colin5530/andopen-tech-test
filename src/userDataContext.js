import React, { useContext, useReducer } from 'react';
import userData from './userData.json';

const UserDataContext = React.createContext();

const initialState = {
  totalUsers: userData.userData,
  visibleUsers: userData.userData,
  spendFilterValue: [0, 5000],
  regionFilterValue: [],
  genderFilterValue: [],
};

const applyFilters = ({totalUsers, spendFilterValue, regionFilterValue, genderFilterValue}) => {
  return totalUsers.filter((user) => {
    return user.spend >= spendFilterValue[0] && user.spend <= spendFilterValue[1];
  }).filter((user) => {
    if (regionFilterValue.length === 0) return true;
    return [...regionFilterValue].includes(user.region);
  }).filter((user) => {
    if (genderFilterValue === 'ALL') return true;
    return genderFilterValue === user.gender;
  });
}

const UserDataReducer = (state, action) => {
  switch (action.type) {
    case 'updateSpendFilter': {
      const filteredUsers = applyFilters(Object.assign({...state}, {spendFilterValue: action.payload}));

      return Object.assign({...state}, {visibleUsers: filteredUsers, spendFilterValue: action.payload})
    }
    case 'updateRegionFilter': {
      const filteredUsers = applyFilters(Object.assign({...state}, {regionFilterValue: action.payload}));

      return Object.assign({...state}, {visibleUsers: filteredUsers, regionFilterValue: action.payload})
    }
    case 'updateGenderFilter': {
      const filteredUsers = applyFilters(Object.assign({...state}, {genderFilterValue: action.payload}));

      return Object.assign({...state}, {visibleUsers: filteredUsers, genderFilterValue: action.payload})
    }
    case 'updateVisibleUsers': {
      return Object.assign({...state}, {visibleUsers: action.payload});
    }
    default: {
      throw new Error(`Unhandled action type: ${action.type}`)
    }
  }
}

const UserDataProvider = ({children}) => {
  const [state, dispatch] = useReducer(UserDataReducer, initialState);
  const value = { state, dispatch };
  return <UserDataContext.Provider value={value}>{children}</UserDataContext.Provider>
}

const useUserData = () => {
  const context = useContext(UserDataContext);
  if (context === undefined) {
    throw new Error('useUserData must be used within a UserDataProvider');
  }
  return {visibleUsers: context.state.visibleUsers, setVisibleUsers: (payload) => context.dispatch({ type: 'updateVisibleUsers', payload})};
}

const useSpendFilter = () => {
  const context = useContext(UserDataContext);
  if (context === undefined) {
    throw new Error('useSpendFilter must be used within a UserDataProvider');
  }
  return {spendFilterValue: context.state.spendFilterValue, setSpendFilterValue: (payload) => context.dispatch({ type: 'updateSpendFilter', payload})};
}

const useRegionFilter = () => {
  const context = useContext(UserDataContext);
  if (context === undefined) {
    throw new Error('useRegionFilter must be used within a UserDataProvider');
  }
  return {regionFilterValue: context.state.regionFilterValue, setRegionFilterValue: (payload) => context.dispatch({ type: 'updateRegionFilter', payload})};
}

const useGenderFilter = () => {
  const context = useContext(UserDataContext);
  if (context === undefined) {
    throw new Error('useGenderFilter must be used within a UserDataProvider');
  }
  return {genderFilterValue: context.state.genderFilterValue, setGenderFilterValue: (payload) => context.dispatch({ type: 'updateGenderFilter', payload})};
}

export {
  UserDataProvider,
  useUserData,
  useSpendFilter,
  useRegionFilter,
  useGenderFilter,
}