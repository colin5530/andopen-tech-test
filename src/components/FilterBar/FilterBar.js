import React, { useState } from 'react';
import { Select, Slider } from 'antd';
import { useRegionFilter, useSpendFilter } from '../../userDataContext';
import './FilterBar.css';

const { Option } = Select;

const FilterBar = () => {
  const className = 'c-FilterBar';
  const {setSpendFilterValue} = useSpendFilter();
  const {setRegionFilterValue} = useRegionFilter();
  const [sliderValue, setSliderValue] = useState([0, 5000]);
  const regions = ['EMEA', 'APAC', 'NAM', 'LATAM'];

  const handleSliderChange = (val) => {
    setSliderValue(val);
  }

  const handleAfterSliderChange = (val) => {
    setSpendFilterValue(val);
  }

  const handleRegionChange = (val) => {
    setRegionFilterValue(val);
  }

  return (
    <div className={`${className}`}>
      <div className={`${className}-slider-container`}>
        <Slider
          range
          min={0}
          max={5000}
          value={[...sliderValue]}
          onChange={handleSliderChange}
          onAfterChange={handleAfterSliderChange}
        />
      </div>
      <div className={`${className}-region-container`}>
        Region
        <Select
          mode="multiple"
          style={{ width: '100%' }}
          placeholder="Region"
          onChange={handleRegionChange}
        >
          {regions.map((region) => (
            <Option key={region}>{region}</Option>
          ))}
        </Select>
      </div>
      <div className={`${className}-gender-container`}>
        Gender
      </div>
    </div>
  )
};

export default FilterBar;