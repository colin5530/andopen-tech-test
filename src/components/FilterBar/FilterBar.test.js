import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import FilterBar from './';
import { UserDataProvider } from '../../userDataContext';

describe('FilterBar', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      <UserDataProvider>
        <FilterBar />
      </UserDataProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders FilterBar with all 3 filters', () => {
    render(
      <UserDataProvider>
        <FilterBar />
      </UserDataProvider>
    );
    
    const spendTitleElement = screen.getByText('Spend:');
    const sliderElements = screen.getAllByRole('slider');

    const regionTitleElement = screen.getAllByText('Region')[0];
    const selectElement = screen.getByRole('combobox');

    const genderTitleElement = screen.getByText('Gender');
    const genderAllElement = screen.getByText('All');
    const genderMaleElement = screen.getByText('Male');
    const genderFemaleElement = screen.getByText('Female');

    expect(spendTitleElement).toBeInTheDocument();
    expect(sliderElements).toHaveLength(2);
    expect(regionTitleElement).toBeInTheDocument();
    expect(selectElement).toBeInTheDocument();
    expect(genderTitleElement).toBeInTheDocument();
    expect(genderAllElement).toBeInTheDocument();
    expect(genderMaleElement).toBeInTheDocument();
    expect(genderFemaleElement).toBeInTheDocument();
  });

  describe('Spend', () => {
    it('renders title, values and slider with 2 handles', () => {
      render(
        <UserDataProvider>
          <FilterBar />
        </UserDataProvider>
      );
      
      const spendTitleElement = screen.getByText('Spend:');
      const sliderElements = screen.getAllByRole('slider');

      const minValueElement = screen.getByTestId('min-value');
      const maxValueElement = screen.getByTestId('max-value');

      expect(spendTitleElement).toBeInTheDocument();
      expect(sliderElements).toHaveLength(2);
      expect(minValueElement).toBeInTheDocument();
      expect(maxValueElement).toBeInTheDocument();
    });

    // it('moving slider handles updates values', () => {
    //   render(
    //     <UserDataProvider>
    //       <FilterBar />
    //     </UserDataProvider>
    //   );

    //   // screen.debug();

    //   let minValueElement = screen.getByTestId('min-value');
    //   let maxValueElement = screen.getByTestId('max-value');

    //   expect(minValueElement).toBeInTheDocument();
    //   expect(minValueElement.textContent).toBe('€0');
    //   expect(maxValueElement).toBeInTheDocument();
    //   expect(maxValueElement.textContent).toBe('€5000');

    //   const sliderElements = screen.getAllByRole('slider');
    //   console.log('slider handle 1', sliderElements[0].style, sliderElements[1].style);

    //   // sliderElements[0].focus()
    //   // fireEvent.click(sliderElements[0]);
    //   fireEvent.focus(sliderElements[0]);
    //   // fireEvent.keyDown(sliderElements[0], { key: 'ArrowRight', code: 'ArrowRight' });
    //   // fireEvent.keyDown(sliderElements[0], { key: 'ArrowRight', code: 'ArrowRight' });
    //   // fireEvent.keyDown(sliderElements[0], { key: 'ArrowRight', code: 'ArrowRight' });

    //   // fireEvent.change(sliderElements[0], { style: { left: '1%' } });

    //   console.log('slider h1 element id', sliderElements[0], sliderElements[0].target);
    //   console.log('document activeelement', document.activeElement, document.activeElement.target);

    //   // console.log('slider handle 1 after', sliderElements[0]);

    //   // screen.debug();

    //   console.log('min-value pre-refresh', minValueElement.textContent);

    //   minValueElement = screen.getByTestId('min-value');

    //   console.log('min-value post-refresh', minValueElement.textContent);
      
    //   expect(minValueElement.textContent).toBe('€3');
    //   expect(maxValueElement.textContent).toBe('€5000');

    //   fireEvent.click(sliderElements[1]);
    //   fireEvent.keyDown(sliderElements[1], { key: 'ArrowLeft', code: 'ArrowLeft' });
    //   fireEvent.keyDown(sliderElements[1], { key: 'ArrowLeft', code: 'ArrowLeft' });
    //   fireEvent.keyDown(sliderElements[1], { key: 'ArrowLeft', code: 'ArrowLeft' });

    //   minValueElement = screen.getByTestId('max-value');
      
    //   expect(minValueElement).toBeInTheDocument();
    //   expect(maxValueElement).toBeInTheDocument();

    // })
  });
});
