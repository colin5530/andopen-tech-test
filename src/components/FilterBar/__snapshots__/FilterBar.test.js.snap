// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterBar renders correctly 1`] = `
<div
  className="c-FilterBar"
>
  <div
    className="c-FilterBar-spend-container"
  >
    <span
      className="c-FilterBar-filter-title"
    >
      Spend: 
    </span>
    €0 - €5000
    <div
      className="c-FilterBar-filter-input"
    >
      <div
        className="ant-slider ant-slider-with-marks c-FilterBar-spend-slider"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        onTouchStart={[Function]}
      >
        <div
          className="ant-slider-rail"
          style={Object {}}
        />
        <div
          className="ant-slider-track ant-slider-track-1"
          style={
            Object {
              "left": "0%",
              "right": "auto",
              "width": "100%",
            }
          }
        />
        <div
          className="ant-slider-step"
        >
          <span
            className="ant-slider-dot ant-slider-dot-active"
            style={
              Object {
                "left": "0%",
              }
            }
          />
          <span
            className="ant-slider-dot ant-slider-dot-active"
            style={
              Object {
                "left": "100%",
              }
            }
          />
        </div>
        <div
          aria-disabled={false}
          aria-valuemax={5000}
          aria-valuemin={0}
          aria-valuenow={0}
          className="ant-slider-handle ant-slider-handle-1"
          onBlur={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          role="slider"
          style={
            Object {
              "left": "0%",
              "right": "auto",
              "transform": "translateX(-50%)",
            }
          }
          tabIndex={0}
        />
        <div
          aria-disabled={false}
          aria-valuemax={5000}
          aria-valuemin={0}
          aria-valuenow={5000}
          className="ant-slider-handle ant-slider-handle-2"
          onBlur={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          role="slider"
          style={
            Object {
              "left": "100%",
              "right": "auto",
              "transform": "translateX(-50%)",
            }
          }
          tabIndex={0}
        />
        <div
          className="ant-slider-mark"
        >
          <span
            className="ant-slider-mark-text ant-slider-mark-text-active"
            onMouseDown={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "left": "0%",
                "msTransform": "translateX(-50%)",
                "transform": "translateX(-50%)",
              }
            }
          >
            €0
          </span>
          <span
            className="ant-slider-mark-text ant-slider-mark-text-active"
            onMouseDown={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "left": "100%",
                "msTransform": "translateX(-50%)",
                "transform": "translateX(-50%)",
              }
            }
          >
            €5000
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    className="c-FilterBar-region-container"
  >
    <span
      className="c-FilterBar-filter-title"
    >
      Region
    </span>
    <div
      className="c-FilterBar-filter-input"
    >
      <div
        className="ant-select c-FilterBar-region-select ant-select-multiple ant-select-show-arrow ant-select-show-search"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        style={
          Object {
            "width": "100%",
          }
        }
      >
        <div
          className="ant-select-selector"
          onClick={[Function]}
          onMouseDown={[Function]}
        >
          <div
            className="ant-select-selection-overflow"
          >
            <div
              className="ant-select-selection-overflow-item ant-select-selection-overflow-item-suffix"
              style={
                Object {
                  "height": undefined,
                  "opacity": 1,
                  "order": undefined,
                  "overflowY": undefined,
                  "pointerEvents": undefined,
                  "position": undefined,
                }
              }
            >
              <div
                className="ant-select-selection-search"
                onBlur={[Function]}
                onFocus={[Function]}
                style={
                  Object {
                    "width": 0,
                  }
                }
              >
                <input
                  aria-activedescendant="undefined_list_0"
                  aria-autocomplete="list"
                  aria-controls="undefined_list"
                  aria-haspopup="listbox"
                  aria-owns="undefined_list"
                  autoComplete="off"
                  className="ant-select-selection-search-input"
                  onChange={[Function]}
                  onCompositionEnd={[Function]}
                  onCompositionStart={[Function]}
                  onKeyDown={[Function]}
                  onMouseDown={[Function]}
                  onPaste={[Function]}
                  readOnly={true}
                  role="combobox"
                  style={
                    Object {
                      "opacity": 0,
                    }
                  }
                  type="search"
                  unselectable="on"
                  value=""
                />
                <span
                  aria-hidden={true}
                  className="ant-select-selection-search-mirror"
                >
                  
                   
                </span>
              </div>
            </div>
          </div>
          <span
            className="ant-select-selection-placeholder"
          >
            Region
          </span>
        </div>
        <span
          aria-hidden={true}
          className="ant-select-arrow"
          onMouseDown={[Function]}
          style={
            Object {
              "WebkitUserSelect": "none",
              "userSelect": "none",
            }
          }
          unselectable="on"
        >
          <span
            aria-label="down"
            className="anticon anticon-down ant-select-suffix"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="down"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
              />
            </svg>
          </span>
        </span>
      </div>
    </div>
  </div>
  <div
    className="c-FilterBar-gender-container"
  >
    <span
      className="c-FilterBar-filter-title"
    >
      Gender
    </span>
    <div
      className="c-FilterBar-filter-input"
    >
      <div
        className="ant-radio-group ant-radio-group-outline c-FilterBar-gender-radios"
      >
        <label
          className="ant-radio-wrapper ant-radio-wrapper-checked"
        >
          <span
            className="ant-radio ant-radio-checked"
            style={Object {}}
          >
            <input
              checked={true}
              className="ant-radio-input"
              name="gender-radio-group"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onKeyUp={[Function]}
              type="radio"
              value="ALL"
            />
            <span
              className="ant-radio-inner"
            />
          </span>
          <span>
            All
          </span>
        </label>
        <label
          className="ant-radio-wrapper"
        >
          <span
            className="ant-radio"
            style={Object {}}
          >
            <input
              checked={false}
              className="ant-radio-input"
              name="gender-radio-group"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onKeyUp={[Function]}
              type="radio"
              value="MALE"
            />
            <span
              className="ant-radio-inner"
            />
          </span>
          <span>
            Male
          </span>
        </label>
        <label
          className="ant-radio-wrapper"
        >
          <span
            className="ant-radio"
            style={Object {}}
          >
            <input
              checked={false}
              className="ant-radio-input"
              name="gender-radio-group"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onKeyUp={[Function]}
              type="radio"
              value="FEMALE"
            />
            <span
              className="ant-radio-inner"
            />
          </span>
          <span>
            Female
          </span>
        </label>
      </div>
    </div>
  </div>
</div>
`;
